esp-hal/src/spi/master.rs

```diff
@@ -2150,29 +2202,95 @@ mod ehal1 {
         }
 
         fn transfer_in_place(&mut self, words: &mut [u8]) -> Result<(), Self::Error> {
-            // Since we have the traits so neatly implemented above, use them!
-            for chunk in words.chunks_mut(FIFO_SIZE) {
-                SpiBus::write(self, chunk)?;
-                SpiBus::flush(self)?;
-                self.driver().read_bytes_from_fifo(chunk)?;
-            }
-            Ok(())
+            self.driver().transfer(words).map(|_| ())
         }
 
         fn flush(&mut self) -> Result<(), Self::Error> {
             self.driver().flush()
         }
     }
+
+    impl<T> SpiBusAsync for Spi<'_, Async, T>
+    where
+        T: Instance,
+    {
+        async fn read(&mut self, words: &mut [u8]) -> Result<(), Self::Error> {
+            // We need to flush because the blocking transfer functions may return while a
+            // transfer is still in progress.
+            self.flush_async().await?;
+            self.driver().read_bytes_async(words).await
+        }
+
+        async fn write(&mut self, words: &[u8]) -> Result<(), Self::Error> {
+            // We need to flush because the blocking transfer functions may return while a
+            // transfer is still in progress.
+            self.flush_async().await?;
+            self.driver().write_bytes_async(words).await
+        }
+
+        async fn transfer(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Self::Error> {
```

Comment:
a specific reason why we don't have the implementation on driver and just delegate like we do for the other functions?

Reply:
Just a bit of cargo culting, this is how the blocking impl is done as well.